### PR TITLE
Adds a drone fabricator and some config options

### PR DESCRIPTION
**Pull request by ghost**
Adds a drone fabricator, ported from bay with much gratitude. Click on it to spawn as a drone if there is a fabricator on the map.

Drone fabricator needs time to recharge after producing each drone. This time is a config option.

There is a max amount of drones config available now too, which counts the number of ALIVE drones on the server and blocks the fabricator from producing more drones if there are too many drones. Default drone cap is 10.

After dying, you need to wait 10 minutes before being able to join as a drone so people don't just deathgasp and immediately go drone when they get killed, so doctors have a fighting chance of actually doing their job.

These fabricators aren't buildable, and need to be placed on the map at the mapper's discretion. I recommend electrical storage for the drone fabricator, or somewhere in engineering maint.

This also basically replaces the need for building drone shells at robotics, which is pretty lame.

### CODE REVIEW

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+	time_last_drone = world.time
+	var/mob/living/simple_animal/drone/D = new /mob/living/simple_animal/drone(get_turf(loc))
+	D.key = player.key
+
+	drone_progress = 0
+
+
+
+/mob/dead/verb/join_as_drone()
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: seems shitty, pretty sure we have an attack_ghost() or ghost_click()

**Comment by RemieRichards** [comment]
*Cough* https://github.com/tgstation/-tg-station/pull/7513 *Cough*

**Comment by Steelpoint** [comment]
Pretty neat either way, would be interesting to make a "Drone's Den" somewhere in maintenance.

SW Maint needs some love.

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+	var/drone_progress = 0
+	var/produce_drones = 1
+	var/time_last_drone = 500
+
+	icon = 'icons/obj/machines/drone_fab.dmi'
+	icon_state = "drone_fab_idle"
+
+/obj/machinery/drone_fabricator/New()
+	..()
```
Comment (Razharas) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: If you dont add anything to new you dont need to override it at all

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+		if(DF.count_drones() >= config.max_maint_drones)
+			src << "\red There are too many active drones in the world for you to spawn."
+			return
+
+		if(DF.drone_progress >= 100)
+			DF.create_drone(src.client)
+			return
+
+	src << "\red There are no available drone spawn points, sorry."
```
Comment (Razharas) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: Remove \red

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		return
+
+	if (!src.stat)
+		return
+
+	if (usr != src)
+		return 0 //something is terribly wrong
+
+	if(jobban_isbanned(src,"Cyborg"))
+		usr << "\red You are banned from playing synthetics and cannot spawn as a drone."
```
Comment (Razharas) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: Remove \red

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		usr << "You have been dead for[pluralcheck] [deathtimeseconds] seconds."
+		usr << "You must wait 10 minutes to respawn as a drone!"
+		return
+
+	for(var/obj/machinery/drone_fabricator/DF in world)
+		if(DF.stat & NOPOWER || !DF.produce_drones)
+			continue
+
+		if(DF.count_drones() >= config.max_maint_drones)
+			src << "\red There are too many active drones in the world for you to spawn."
```
Comment (Razharas) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: Remove \red

**Comment by AlexCrimson** [comment]
When you click "join as drone" does it give you the popup that warns you about the whole "makes you unable to be cloned" thing?

Is it...free? Is there a cooldown? Could 10 separate ghosts join as drones instantly? Would afk/ghosted drones count towards the cap?

I like it though. Hell, eventually could you even make it a constructable machine that requires a board from R&D?

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+/mob/dead/verb/join_as_drone()
+
+	set category = "Ghost"
+	set name = "Join As Drone"
+	set desc = "If there is a powered, enabled fabricator in the game world with a prepared chassis, join as a maintenance drone."
+
+
+	if(ticker.current_state < GAME_STATE_PLAYING)
+		src << "\red The game hasn't started yet!"
```
Comment (Razharas) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: You know the drill

**Comment by Razharas** [comment]
Also yes currently this is neither on the map nor buildable so looks like adminonly feature

**Comment by AlexCrimson** [comment]
Steelpoint did a redesign of Robotics, so maybe he could add it in.

**Comment by Steelpoint** [comment]
Steelpoint is doing one yes.

I'll investigate it if this is accepted and merged, I could possibly fit it in the mechbay depending on its size, that or I can toss it in maintenance somewhere.

**Comment by Razharas** [comment]
Also having 2 drone factories of different kind that work kinda identically is strange, settle this down with perakp

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm (hunk 2 above)
Comment (paprka) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: Welcome to copypaste

**Comment by paprka** [comment]
It's near engineering on bay, maybe put it where the old incinerator used to be and remove the ghetto mix room that is there.

**Comment by AlexCrimson** [comment]
Wouldnt it be best to just give the AI an ability to hack the Drone Fab and turn it into whatever Perakp coded? The only issue would be the metaknowledge of it always being in the same location, so people could easily destroy it.

**Comment by paprka** [comment]
true, would be cool

**Comment by Fox-McCloud** [comment]
If added, I also suggest adding a "join as drone" to the player options. Plenty of people will want to play as drone, but not want to wait 10 minutes just to do so. It also prevent slot waste of people who join, insta-suicide just to get drone.

**Comment by paprka** [comment]
Drones shouldn't really be a roundstart thing to be honest

**Comment by Fox-McCloud** [comment]
Yeah, they can, but in that point, the 10 minute wait just seems like a "screw you" think rather than a real barrier to prevent abuses. It also means that they're available sooner for traitors to potentially emag.

**Comment by JJRcop** [comment]
@Fox-McCloud Drones cannot be emagged.

**Comment by paprka** [comment]
@Perakp what are your thoughts on this and integrating your PR with this fabricator thing?

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm (hunk 4 above)
Comment (Alek2ander) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: Also, drones use the pAI jobban, not the borg one.

**Comment by Alek2ander** [comment]
Can this maybe be constructable? I don't like the idea of 10 drones spawning early in the round.

**Comment by Saegrimr** [comment]
10 minutes seems harsh for the dead, however drones that get killed should have some kind of timer to keep them from jumping right into a new drone and going off to get killed by usually the exact same thing again.

File: code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm (hunk 4 above)
Comment (paprka) [inline code/modules/mob/living/simple_animal/friendly/drone_fabricator.dm]: Nice catch alek

**Comment by Jordie0608** [comment]
Changelog.

**Comment by paprka** [comment]
- Removed drone shells from exo fabs
- Added a drone fabricator to box
- Fixed some \red and a useless new() override

**Comment by Cheridan** [comment]
I'm kinda agreeing with @RemieRichards about the verb. If a player is brand-spanking-new he probably shouldn't be playing Drones yet anyway, because of their gameplay restrictions will likely get him banned.

Having them work off a ghost's click is less UI clutter and gives it a (very small) cost of entry.

**Comment by paprka** [comment]
Alright, I'll convert it over to that then.

**Comment by paprka** [comment]
No more verb, you now click the machine as a ghost to spawn as a drone.

@Cheridan

**Comment by Cheridan** [comment]
SOON

**Comment by paprka** [comment]
Made a maint area for drones

ss 2015-02-18 at 10 53 36

It's shitty and cramped but it works and a mapper can improve on it later, better than just putting it in construction for now @Cheridan

**Comment by Miauw62** [comment]
what maintainer is handling this, @Razharas @Cheridan ???

**Comment by paprka** [comment]
Cheridan but he's AWOL

**Comment by paprka** [comment]
WELP, SO MUCH FOR DRONE AREA IN MAINT

CONFLICTS FIXED!!!!!!!!!!!!!!!!!

**Comment by paprka** [comment]
Oh well so much for this PR

**Comment by AlexCrimson** [comment]
Damn. Real shame.

**Pull request closed without merging**